DataProcessing/CoinGeckoDataDownloader.cs

```diff
@@ -0,0 +1,383 @@
+/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using Newtonsoft.Json;
+using QuantConnect.DataSource;
+using QuantConnect.Logging;
+using QuantConnect.Securities;
+using QuantConnect.Util;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace QuantConnect.DataProcessing
+{
+    /// <summary>
+    /// CoinGeckoUniverseDataDownloader implementation.
+    /// </summary>
+    public class CoinGeckoUniverseDataDownloader : IDisposable
+    {
+        public const string VendorName = "coingecko";
+        
+        private readonly string _destinationFolder;
+        private readonly string _processedFolder;
+        private readonly string _universeFolder;
+        private readonly string _processedUniverseFolder;
+        private readonly int _maxRetries = 5;
+        private readonly HttpClient _client;
+
+        /// <summary>
+        /// Control the rate of download per unit of time.
+        /// </summary>
+        private readonly RateGate _indexGate;
+
+        /// <summary>
+        /// Creates a new instance of <see cref="CoinGeckoUniverseDataDownloader"/>
+        /// </summary>
+        /// <param name="destinationFolder">The folder where the data will be saved</param>
+        /// <param name="apiKey">The Vendor API key</param>
+        public CoinGeckoUniverseDataDownloader(string destinationFolder, string processedFolder, string apiKey = null)
```

Comment:
>  string apiKey = null

Isn't used, should we just remove it then?
Missing xml docs for `processedFolder`